amethyst_ui/src/selection.rs

```diff
@@ -0,0 +1,268 @@
+    #[new(default)]
+    phantom: PhantomData<(G, AX, AC)>,
+}
+
+impl<'a, G, AX, AC> System<'a> for SelectionMouseSystem<G, AX, AC>
+where
+    G: Send + Sync + 'static + PartialEq,
+    AX: Hash + Eq + Clone + Send + Sync + 'static,
+    AC: Hash + Eq + Clone + Send + Sync + 'static,
+{
+    type SystemData = (
+        Read<'a, EventChannel<UiEvent>>,
+        Read<'a, CachedSelectionOrder>,
+        WriteStorage<'a, Selected>,
+        ReadStorage<'a, Selectable<G>>,
+        Read<'a, InputHandler<AX, AC>>,
+    );
+    fn run(
+        &mut self,
+        (ui_events, cached, mut selecteds, selectables, input_handler): Self::SystemData,
+    ) {
+        let shift = input_handler.key_is_down(VirtualKeyCode::LShift)
+            || input_handler.key_is_down(VirtualKeyCode::RShift);
+        let ctrl = input_handler.key_is_down(VirtualKeyCode::LControl)
+            || input_handler.key_is_down(VirtualKeyCode::RControl);
+
+        // Add clicked elements to clicked buffer
+        for ev in ui_events.read(self.ui_reader_id.as_mut().unwrap()) {
+            match ev.event_type {
+                UiEventType::ClickStart => {
+                    // Ignore events from elements removed between the event emission and now.
+                    if selectables.get(ev.target).is_some() {
+                        let clicked = ev.target;
+
+                        // Inside of the loop because its possible that the user clicks two times in a frame while pressing shift.
+                        let highest = cached.highest_order_selected_index(&selecteds);
+
+                        if let Some(highest) = highest {
+                            // Safe unwraps, we just got those values from the cache.
+
+                            let (highest_is_select, auto_multi_select) = {
+                                let highest_multi_select_group = &selectables
+                                    .get(cached.cache.get(highest).unwrap().1)
+                                    .unwrap()
+                                    .multi_select_group;
+
+                                let (target_multi_select_group, auto_multi_select) = {
+                                    // Safe unwrap because when filing the buffer we checked that the component still exist on the entity.
+                                    let target_selectable = selectables.get(clicked).unwrap();
+                                    (
+                                        &target_selectable.multi_select_group,
+                                        target_selectable.auto_multi_select,
+                                    )
+                                };
+                                (
+                                    highest_multi_select_group == target_multi_select_group,
+                                    auto_multi_select,
+                                )
+                            };
+
+                            if highest_is_select {
```

Comment:
I'd turn around if and else here, makes it way easier to read since you don't have to scroll to get to the else

Reply:
Its that way to reflect the algorithm written in comment at the top.

Reply:
I see, that's fair